File: labs/cpp/invcount.cpp
#include <iostream>
#include <vector>
using namespace std;

long long merge_count(vector<int>& a, int l, int r, vector<int>& tmp) {
    if (r - l <= 1) return 0;
    int m = l + (r - l) / 2;
    long long c = 0;
    c += merge_count(a, l, m, tmp);
    c += merge_count(a, m, r, tmp);
    int i = l, j = m, k = l;
    while (i < m && j < r) {
        if (a[i] <= a[j]) tmp[k++] = a[i++];
        else {
            tmp[k++] = a[j++];
            c += m - i;
        }
    }
    while (i < m) tmp[k++] = a[i++];
    while (j < r) tmp[k++] = a[j++];
    for (int t = l; t < r; ++t) a[t] = tmp[t];
    return c;
}

long long inversion_count(vector<int> a) {
    vector<int> tmp(a.size());
    return merge_count(a, 0, (int)a.size(), tmp);
}

int main() {
    vector<int> arr = {2, 4, 1, 3, 5};
    cout << inversion_count(arr) << "\n";
    return 0;
}

File: labs/py/invcount.py
from typing import List, Tuple
import random

def count_and_sort(a: List[int]) -> Tuple[List[int], int]:
    n = len(a)
    if n <= 1:
        return a[:], 0
    m = n // 2
    left, cL = count_and_sort(a[:m])
    right, cR = count_and_sort(a[m:])
    merged = []
    i = 0
    j = 0
    cross = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            cross += len(left) - i
            j += 1
    merged.extend(left[i:])
    merged.extend(right[j:])
    return merged, cL + cR + cross

def inversion_count(a: List[int]) -> int:
    _, c = count_and_sort(a)
    return c

def brute_inversions(a: List[int]) -> int:
    n = len(a)
    c = 0
    for i in range(n):
        for j in range(i + 1, n):
            if a[i] > a[j]:
                c += 1
    return c

if __name__ == "__main__":
    for n in [5, 10, 50, 100]:
        arr = [random.randint(0, 999) for _ in range(n)]
        fast = inversion_count(arr)
        brute = brute_inversions(arr)
        print(n, fast, brute, "OK" if fast == brute else "MISMATCH")
